libretroshare/src/grouter/p3grouter.cc

```diff
@@ -1485,132 +1536,204 @@ bool p3GRouter::sliceDataItem(RsGRouterAbstractMsgItem *item,std::list<RsGRouter
 
-        if(it->second.routing_flags & GRouterRoutingInfo::ROUTING_FLAGS_IS_ORIGIN)
-        {
-#ifdef GROUTER_DEBUG
-            std::cerr << "  message is at origin. Setting message transmission to DONE" << std::endl;
-#endif
-            it->second.data_status  = RS_GROUTER_DATA_STATUS_DONE;
+		// check hash.
+		if(receiptItem->data_hash != it->second.item_hash)
+		{
+			RsWarn() << __PRETTY_FUNCTION__ << " receipt hash doesn't match "
+			         << " dropping" << std::endl;
+			return;
+		}
 
-            if(locked_getLocallyRegisteredClientFromServiceId(it->second.client_id,client_service))
-                mid = it->first ;
-            else
-            {
-                mid = 0 ;
-                std::cerr << " ERROR: cannot retrieve service ID for message " << std::hex << it->first << std::dec << std::endl;
-            }
-        }
-        else
-        {
-#ifdef GROUTER_DEBUG
-            std::cerr << "  message is not at origin. Setting message transmission to RECEIPT_OK" << std::endl;
-#endif
-            it->second.data_status  = RS_GROUTER_DATA_STATUS_RECEIPT_OK;
-            it->second.receipt_item = receipt_item->duplicate() ;
-        }
+		/* If the receipt is of a message we sent, look for the service who sent
+		 * the original message and prepare pointer to notify message has been
+		 * received, if we can't find the service print an error as something
+		 * fishy is happening.*/
+		if( it->second.routing_flags &
+		        GRouterRoutingInfo::ROUTING_FLAGS_IS_ORIGIN )
+		{
 
-        changed = true ;
-    }
+			it->second.data_status  = RS_GROUTER_DATA_STATUS_DONE;
 
-    if(mid != 0)
-    {
-#ifdef GROUTER_DEBUG
-        std::cerr << "  notifying client " << (void*)client_service << " that msg " << std::hex << mid << std::dec << " was received." << std::endl;
-#endif
-        client_service->notifyDataStatus(mid, signer_id, GROUTER_CLIENT_SERVICE_DATA_STATUS_RECEIVED) ;
-    }
+			if(!locked_getLocallyRegisteredClientFromServiceId(
+			            clientServiceType, client_service ))
+				RsErr() << __PRETTY_FUNCTION__ << " cannot retrieve client "
+				        << "service: " << it->second.client_id
+				        << " instance for receipt to own message "
+				        << std::hex << it->first << std::dec << std::endl;
+		}
+		/* If the receipt is for a known message that is not from us mark the
+		 * message as received and, associate the receipt to it
+		 * TODO: is this enough to ensure the receipt is forwarded to
+		 * potentially interested peers? */
+		else
+		{
+			Dbg2() << __PRETTY_FUNCTION__ << "  message is not at origin. "
+			       << "Setting message transmission to RECEIPT_OK" << std::endl;
 
-    // also note the incoming route in the routing matrix
+			it->second.data_status  = RS_GROUTER_DATA_STATUS_RECEIPT_OK;
+			it->second.receipt_item = receiptItem.release();
+			// no need of dluplicate() just release unique_ptr
+		}
 
-    if(!mTurtle->isTurtlePeer(receipt_item->PeerId()))
```

Comment:
Why is this removed?

Reply:
It hasn't been removes it, has been moved at line 1649, because this method has been changed a plain diff won't help much